backend/tilavarauspalvelu/services/first_reservable_time/first_reservable_time_reservation_unit_helper.py

```diff
@@ -129,6 +130,27 @@ def calculate_first_reservable_time(self) -> ReservableTimeOutput:
 
         return ReservableTimeOutput(is_closed=self.is_reservation_unit_closed, first_reservable_time=None)
 
+    def get_access_type_for_date(
+        self,
+        is_closed: bool,  # noqa: FBT001
+        first_reservable_time: datetime.datetime | None,
+    ) -> AccessType | None:
+        if is_closed or first_reservable_time is None:
+            access_type_date = self.parent.filter_date_start
+        else:
+            access_type_date = first_reservable_time.date()
+
+        # Loop through the access types in REVERSE order to find the access type active at the first reservable time
+        access_types: list[ReservationUnitAccessType] = sorted(
+            self.reservation_unit.access_types.all(),
+            key=lambda x: x.begin_date,
+            reverse=True,
+        )
```

Comment:
I think you could do this in the prefetch queryset? Not that there will be so many access types that it matters much.